core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go

```diff
@@ -123,6 +124,86 @@ func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]
 	return bts, nil
 }
 
+// should accept/transmit reports from plugin will call extract function
+func (enc EVMAutomationEncoder21) Extract(report []byte) ([]ocr2keepers.ReportedUpkeep, error) {
+	m := make(map[string]interface{})
+	if err := unpackIntoMapFn(m, report); err != nil {
+		return nil, err
+	}
+
+	for _, key := range mKeys {
+		if _, ok := m[key]; !ok {
+			return nil, fmt.Errorf("decoding error: %s missing from struct", key)
+		}
+	}
+
+	var (
+		res       []ocr2keepers.ReportedUpkeep
+		ok        bool
+		upkeepIds []*big.Int
+		performs  [][]byte
+		// gasLimits []*big.Int // TODO
+		//wei  *big.Int
+		//link *big.Int
+	)
+
+	if upkeepIds, ok = m[mKeys[2]].([]*big.Int); !ok {
+		return res, fmt.Errorf("upkeep ids of incorrect type in report")
+	}
+
+	// TODO: a type assertion on `wrappedTrigger` did not work, even with the
+	// exact same struct definition as what follows. reflect was used to get the
+	// struct definition. not sure yet how to clean this up.
+	// ex:
+	// t := reflect.TypeOf(rawPerforms)
+	// fmt.Printf("%v\n", t)
+	triggers, ok := m[mKeys[4]].([]struct {
```

Comment:
@amirylm you will add 2 more fields here?

Reply:
You should be able to use the struct from the contract to derive the fields to use. Looks like the field names are different. I will need to change mine too.

```
  struct LogTrigger {
    bytes32 txHash;
    uint32 logIndex;
    uint32 blockNum;
    bytes32 blockHash;
  }
```